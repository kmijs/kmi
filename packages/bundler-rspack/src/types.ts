import type {
  PluginItem as BabelPlugin,
  TransformOptions as BabelTransformOptions,
} from '@kmijs/bundler-compiled/compiled/babel/core'
import type { KmiTarget } from '@kmijs/bundler-shared'
import { Env, type IBundlerConfig } from '@kmijs/bundler-shared-config'
import type {
  LightningcssLoaderOptions,
  SwcLoaderOptions,
} from '@kmijs/bundler-shared/rspack'

export enum Transpiler {
  babel = 'babel',
  swc = 'swc',
  none = 'none',
}

export { Env }

export interface DeadCodeParams {
  patterns?: string[]
  exclude?: string[]
  failOnHint?: boolean
  detectUnusedFiles?: boolean
  detectUnusedExport?: boolean
  context?: string
}

export type ExtraSwcPlugins = Array<[string, any]>

export type ModifySwcLoaderOptions = (
  options: SwcLoaderOptions,
  opts: {
    env: Env
    target: KmiTarget
    isServer: boolean
    isWebWorker: boolean
  },
) => SwcLoaderOptions

export interface IConfig extends IBundlerConfig {
  cssMinifierOptions?: { [key: string]: any }
  depTranspiler?: `${Transpiler}`
  srcTranspiler?: `${Transpiler}`
  srcTranspilerOptions?: Record<string, any>
  svgr?: { [key: string]: any }
  svgo?: { [key: string]: any } | false
  swc?: SwcLoaderOptions
  extraSwcPlugins?: ExtraSwcPlugins
  modifySwcLoaderOptions?: ModifySwcLoaderOptions
  lightningcssLoader?: boolean | LightningcssLoaderOptions
  [key: string]: any
}

export type BabelLoaderOptions = BabelTransformOptions & {
  /**
   * When set, the given directory will be used to cache the results of the loader.
   */
  cacheDirectory?: string | boolean
  /**
   * Can be set to a custom value to force cache busting if the identifier changes.
   */
  cacheIdentifier?: string
  /**
   * When set, each Babel transform output will be compressed with Gzip.
   */
  cacheCompression?: boolean
  /**
   * The path of a module that exports a custom callback.
   */
  customize?: string | null
  /**
   * Takes an array of context function names. E.g.
   */
  metadataSubscribers?: string[]
}

type OneOrMany<T> = T | T[]
export type ConfigChainWithContext<T, Ctx> = OneOrMany<
  T | ((config: T, ctx: Ctx) => T | void)
>

export interface PresetTypescriptOptions {
  ignoreExtensions?: boolean
  allowDeclareFields?: boolean
  allowNamespaces?: boolean
  disallowAmbiguousJSXLike?: boolean
  jsxPragma?: string
  jsxPragmaFrag?: string
  onlyRemoveTypeImports?: boolean
  optimizeConstEnums?: boolean
  rewriteImportExtensions?: boolean

  allExtensions?: boolean
  isTSX?: boolean
}

export type BabelConfigUtils = {
  addPlugins: (plugins: BabelPlugin[]) => void
  addPresets: (presets: BabelPlugin[]) => void
  removePlugins: (plugins: string | string[]) => void
  removePresets: (presets: string | string[]) => void
  modifyPresetTypescriptOptions: (options: PresetTypescriptOptions) => void
  modifyPluginOptions: (
    pluginName: string,
    options: Record<string, any>,
  ) => void
  isWebWorker?: boolean
  isServer?: boolean
}

export interface IExtraRspackOpts {
  extraSwcPlugins?: ExtraSwcPlugins
  modifySwcLoaderOptions?: ModifySwcLoaderOptions
}
