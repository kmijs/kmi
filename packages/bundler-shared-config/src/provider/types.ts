import type { BabelPlugin } from '@kmijs/bundler-compiled'
import type { KmiTarget } from '@kmijs/bundler-shared'
import type {
  Application,
  RequestHandler,
} from '@kmijs/bundler-shared/compiled/express'
import type {
  Compiler,
  Configuration,
  MultiCompiler,
} from '@kmijs/bundler-shared/rspack'
import type Config from '@kmijs/bundler-shared/rspack-chain'
import type { Env, IUserConfig } from '../types'

export interface IBaseConfigOpts<U extends IUserConfig = IUserConfig> {
  cwd: string
  rootDir?: string
  entry: Record<string, string>
  config: U
  babelPreset?: BabelPlugin[]
  beforeBabelPlugins?: BabelPlugin[]
  beforeBabelPresets?: BabelPlugin[]
  extraBabelPlugins?: BabelPlugin[]
  extraBabelPresets?: BabelPlugin[]
  disableCopy?: boolean
  cache?: any
  pkg?: any
  chainWebpack?: (config: any, opts: any) => void
  modifyWebpackConfig?: (config: any) => any
  /**
   * onBeforeCreateCompiler 是在创建底层 Compiler 实例前触发的回调函数
   * 你可以通过 bundlerConfigs 参数获取到 Rspack 配置数组，数组中可能包含一份或多份 Rspack(Webpack) 配置，这取决于是否配置了多个 environments。
   * @param args
   * @returns
   */
  onBeforeCreateCompiler?: (...args: any[]) => void
  /**
   * onAfterCreateCompiler 是在创建底层 Compiler 实例后触发的回调函数
   * 你可以通过 compiler 参数获取到 Compiler 实例对象
   * @param args
   * @returns
   */
  onAfterCreateCompiler?: (...args: any[]) => void
  /**
   * 构建目标
   * @default 'web'
   * web: 浏览器
   * node: 服务端
   * web-worker: 浏览器 worker
   */
  target?: KmiTarget
}

export interface IBaseServerOpts {
  onDevCompileDone?: (opts: {
    stats: any
    isFirstCompile: boolean
    time: number
  }) => void
  onProgress?: ({ progresses }: { progresses: any[] }) => void
  onAfterStartDevServer?: (opts: {
    protocol: string
    port: number
    host: string
  }) => Promise<void>
  onBeforeMiddleware?: (app: Application) => void
  beforeMiddlewares?: RequestHandler[]
  afterMiddlewares?: RequestHandler[]
  port?: number
  host?: string
  ip?: string
  hmr: boolean
}

export interface IDevOpts<U extends IUserConfig = IUserConfig>
  extends IBaseConfigOpts<U>,
    Omit<IBaseServerOpts, 'hmr'> {}

export interface IBuildOpts<U extends IUserConfig = IUserConfig>
  extends IBaseConfigOpts<U> {
  onBuildComplete?: (...args: any[]) => void
  clean?: boolean
  watch?: boolean
}

export type BundlerType = 'webpack' | 'rspack'

export interface IBaseBundlerConfigOpts<U extends IUserConfig = IUserConfig> {
  cwd: string
  rootDir?: string
  env: Env
  entry: Record<string, string>
  extraBabelPresets?: BabelPlugin[]
  extraBabelPlugins?: BabelPlugin[]
  extraBabelIncludes?: Array<string | RegExp>
  babelPreset?: BabelPlugin[]
  chainWebpack?: (config: Config, opts: any) => void
  modifyWebpackConfig?: (config: Configuration, opts: any) => Configuration
  hash?: boolean
  hmr?: boolean
  staticPathPrefix?: string
  userConfig: U
  analyze?: any
  name?: string
  cache?: {
    absNodeModulesPath?: string
    buildDependencies?: string[]
    cacheDirectory?: string
  }
  pkg?: Record<string, any>
  disableCopy?: boolean
  host?: string
  port?: number
  environment?: string
  /**
   * onBeforeCreateCompiler 是在创建底层 Compiler 实例前触发的回调函数
   * 你可以通过 bundlerConfigs 参数获取到 Rspack 配置数组，数组中可能包含一份或多份 Rspack(Webpack) 配置，这取决于是否配置了多个 environments。
   * @param args
   * @returns
   */
  onBeforeCreateCompiler?: (...args: any[]) => void
  /**
   * onAfterCreateCompiler 是在创建底层 Compiler 实例后触发的回调函数
   * 你可以通过 compiler 参数获取到 Compiler 实例对象
   * @param args
   * @returns
   */
  onAfterCreateCompiler?: (...args: any[]) => void
  /**
   * 构建目标
   * @default 'web'
   * web: 浏览器
   * node: 服务端
   * web-worker: 浏览器 worker
   */
  target: KmiTarget
}

export interface IBaseCreateServerOpts<U extends IUserConfig = IUserConfig>
  extends Omit<IBaseBundlerConfigOpts<U>, 'hmr'>,
    IBaseServerOpts {
  bundlerConfigs: Configuration | Configuration[]
  bundlerType: BundlerType
  createCompiler: (opts: any) => Compiler | MultiCompiler
}
