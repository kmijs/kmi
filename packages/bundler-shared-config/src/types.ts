import type {
  BabelPlugin,
  BundleAnalyzerOptions,
  CSSMinifier,
  JSMinifier,
  ManifestOptions,
} from '@kmijs/bundler-compiled'
import type {
  HttpsServerOptions,
  KmiTarget,
  ProxyOptions,
} from '@kmijs/bundler-shared'
import type {
  Configuration,
  LightningcssLoaderOptions,
  LoaderContext,
  RuleSetCondition,
  rspack,
} from '@kmijs/bundler-shared/rspack'
import type Config from '@kmijs/bundler-shared/rspack-chain'

export interface BundlerPluginInstance {
  [index: string]: any
  apply: (compiler: any) => void
}

export interface Bundler {
  BannerPlugin: BundlerPluginInstance
  DefinePlugin: BundlerPluginInstance
  IgnorePlugin: BundlerPluginInstance
  ProvidePlugin: BundlerPluginInstance
  HotModuleReplacementPlugin: BundlerPluginInstance
  CopyPlugin: BundlerPluginInstance
  ManifestPlugin: BundlerPluginInstance
  TsCheckerPlugin: BundlerPluginInstance
  CssExtractPlugin: BundlerPluginInstance
  RuntimePublicPathPlugin: BundlerPluginInstance
}

export enum Env {
  development = 'development',
  production = 'production',
}

type WebpackConfig = Required<Configuration>

export interface ICopy {
  from: string
  to: string
}

type ModifyWebpackOpts = {
  env: Env
  webpack?: typeof rspack
  rspack?: typeof rspack
}

export type IChainWebpack = (config: Config, opts: ModifyWebpackOpts) => void
export type IModifyWebpackConfig = (
  config: Configuration,
  opts: ModifyWebpackOpts,
) => Configuration

export interface IUserConfig {
  define?: { [key: string]: any }
  https?: HttpsServerOptions
  extraBabelPlugins?: BabelPlugin[]
  extraBabelPresets?: BabelPlugin[]
  extraBabelIncludes?: Array<string | RegExp>
  hash?: boolean
  jsMinifier?: `${JSMinifier}`
  jsMinifierOptions?: { [key: string]: any }
  proxy?: { [key: string]: ProxyOptions } | ProxyOptions[]
  styleLoader?: { [key: string]: any }
  targets?: { [key: string]: any }
  writeToDisk?: boolean
  esbuildMinifyIIFE?: boolean
  base?: string
  devtool?: Config.DevTool
  alias?: Record<string, string>
  externals?: WebpackConfig['externals']
  javascriptExportsPresence?: boolean
  outputPath?: string | OutputPathConfig
  publicPath?: string
  esm?: { [key: string]: any }
  inlineLimit?: number | InlineLimitConfig
  mdx?: Record<string, any>
  ignoreMomentLocale?: boolean
  copy?: ICopy[] | string[]
  analyze?: BundleAnalyzerOptions
  manifest?: Partial<ManifestOptions>
  // TODO 改成真正的类型
  forkTSChecker?: Record<string, any>
  /**
   *  设置构建产物的名称
   *  在生产模式构建后，Kmi 会自动在文件名中间添加 hash 值，如果不需要添加，可以将 hash 设置为 false 来禁用该行为。
   */
  filename?: FilenameConfig
  /**
   * 用于控制是否输出图片、字体、音频、视频等静态资源。
   */
  emitAssets?: boolean
  /**
   * 包含应该被视为静态资源的额外文件。
   * @default undefined
   */
  assetsInclude?: RuleSetCondition

  /**
   * 用户 启用 rspack 相关能力, 这里不是 rspack 配置
   */
  rspack?: Record<string, any>

  lightningcssLoader?: boolean | LightningcssLoaderOptions

  /**
   * less loader 主题配置
   */
  theme?: Record<string, any>

  lessLoader?:
    | Record<string, any>
    | ((loaderContext: LoaderContext) => Record<string, any>)
  sassLoader?: Record<string, any>
  stylusLoader?: Record<string, any>
  cssExtractLoader?: Record<string, any>
  cssLoaderModules?: Record<string, any>
  cssLoader?: Record<string, any>
  normalCSSLoaderModules?: Record<string, any>
  extraPostCSSPlugins?: any[]
  /**
   * 是否启用 css modules
   */
  autoCSSModules?: boolean
  autoprefixer?: Record<string, any>
  postcssLoader?: Record<string, any>
  /**
   * 按照顺序解析模块，例如 require('./index')，会依次尝试解析 './index.js'、'./index.json'
   */
  extensions?: string[]
  /**
   * 压缩 css 的工具
   */
  cssMinifier?: `${CSSMinifier}`
  /**
   * 压缩 css 的工具配置
   */
  cssMinifierOptions?: Record<string, any>
  /**
   * 启用运行时 publicPath，开启后会使用 window.publicPath 作为资源动态加载的起始路径
   */
  runtimePublicPath?: Record<string, any> | boolean
  /**
   * 用户提供的链式配置
   */
  chainWebpack?: IChainWebpack
  /**
   * 用户提供的配置修改器
   */
  modifyWebpackConfig?: IModifyWebpackConfig

  /**
   * 开发服务器配置
   */
  server?: {
    open?: string | string[] | boolean
    host?: string
    port?: number
  }
}

export interface SharedConfigOptions {
  name: string
  bundler: Bundler
  cwd: string
  env: Env
  config: Config
  /**
   * 用户配置
   */
  userConfig: IUserConfig
  bundlerType: 'webpack' | 'rspack'
  /**
   * 资源是否带 hash 后缀
   */
  useHash: boolean
  /**
   * 静态资源前缀
   */
  staticPathPrefix: string
  /**
   * 入口 entry
   */
  entry: Record<string, string>
  /**
   * 是否启用热更新
   */
  hmr?: boolean
  /**
   * 禁用文件 copy
   */
  disableCopy?: boolean

  /**
   * 浏览器兼容性配置
   */
  browsers: string[]

  /**
   * 构建目标
   * @default 'web'
   * web: 浏览器
   * node: 服务端
   * web-worker: 浏览器 worker
   */
  target: KmiTarget
}

export type InlineLimitConfig = {
  /**
   * SVG 图片的 data URI 限制大小
   * @default 4096
   */
  svg?: number
  /**
   * 字体文件的 data URI 限制大小
   * @default 4096
   */
  font?: number
  /**
   * 非 SVG 图片的 data URI 限制大小
   * @default 4096
   */
  image?: number
  /**
   * 媒体资源(如视频)的 data URI 限制大小
   * @default 0
   */
  media?: number
  /**
   * 其他静态资源的 data URI 限制大小
   * @default 4096
   */
  assets?: number
}

export type OutputPathConfig = {
  /**
   * 所有文件的根目录
   * @default 'dist'
   **/
  root?: string
  /**
   * JavaScript 文件的输出目录
   * @default 'static/js'
   */
  js?: string
  /**
   * CSS 文件的输出目录
   * @default 'static/css'
   */
  css?: string
  /**
   * SVG 图片的输出目录
   * @default 'static/svg'
   */
  svg?: string
  /**
   * 字体文件的输出目录
   * @default 'static/font'
   */
  font?: string
  /**
   * Wasm 文件的输出目录
   * @default 'static/wasm'
   */
  wasm?: string
  /**
   * 非 SVG 图片的输出目录
   * @default 'static/image'
   */
  image?: string
  /**
   * 媒体资源的输出目录，如视频等
   * @default 'static/media'
   */
  media?: string
  /**
   * 除了上述资源(图片、SVG、字体、HTML、WASM等)以外的资源输出目录
   * @default 'static/assets'
   */
  assets?: string
}

export type FilenameConfig = {
  /**
   * JavaScript 文件的命名规则
   * @default
   * - dev: '[name].js'
   * - prod: '[name].[contenthash:8].js'
   */
  js?: NonNullable<Configuration['output']>['filename']
  /**
   * 异步 JavaScript 文件的命名规则
   * @default
   * - dev: '[name].async.js'
   * - prod: '[name].[contenthash:8].async.js'
   */
  jsAsync?: NonNullable<Configuration['output']>['filename']
  /**
   * CSS 文件的命名规则
   * @default
   * - dev: '[name].css'
   * - prod: '[name].[contenthash:8].css'
   */
  css?: NonNullable<Configuration['output']>['cssFilename']
  /**
   * 异步 CSS 文件的命名规则
   * @default
   * - dev: '[name].async.css'
   * - prod: '[name].[contenthash:8].async.css'
   */
  cssAsync?: NonNullable<Configuration['output']>['cssFilename']
  /**
   * SVG 图片的命名规则
   * @default '[name].[contenthash:8].svg'
   */
  svg?: string
  /**
   * HTML 文件的命名规则
   * @default `[name].html`
   */
  html?: string
  /**
   * 字体文件的命名规则
   * @default '[name].[contenthash:8][ext]'
   */
  font?: string
  /**
   * 非 SVG 图片的命名规则
   * @default '[name].[contenthash:8][ext]'
   */
  image?: string
  /**
   * 媒体资源的命名规则，如视频等
   * @default '[name].[contenthash:8][ext]'
   */
  media?: string
  /**
   * 其他资源的命名规则，除了上述资源(图片、SVG、字体、HTML、WASM等)以外的文件
   * @default '[name].[contenthash:8][ext]'
   */
  assets?: string
}
