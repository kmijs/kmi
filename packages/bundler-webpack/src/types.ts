import type { IBundlerConfig } from '@kmijs/bundler-shared-config'
import type { TransformOptions as EsbuildOptions } from '@kmijs/bundler-shared/esbuild'

export { Env } from '@kmijs/bundler-shared-config'

export enum Transpiler {
  babel = 'babel',
  esbuild = 'esbuild',
  none = 'none',
}

export interface ICopy {
  from: string
  to: string
}

export interface DeadCodeParams {
  patterns?: string[]
  exclude?: string[]
  failOnHint?: boolean
  detectUnusedFiles?: boolean
  detectUnusedExport?: boolean
  context?: string
}

export interface IConfig extends IBundlerConfig {
  cssMinifierOptions?: { [key: string]: any }
  depTranspiler?: `${Transpiler}`
  srcTranspiler?: `${Transpiler}`
  srcTranspilerOptions?: ISrcTranspilerOpts
  svgr?: { [key: string]: any }
  svgo?: { [key: string]: any } | false
  babelLoaderCustomize?: string
  esbuildMinifyIIFE?: boolean
  checkDepCssModules?: boolean
  [key: string]: any
}

export interface ISrcTranspilerOpts {
  esbuild?: Partial<EsbuildOptions>
}

export interface ISwcPluginOpts {
  enableAutoCssModulesPlugin?: boolean
}
