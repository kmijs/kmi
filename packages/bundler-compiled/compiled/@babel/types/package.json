{"name":"@babel/types","version":"7.23.6","author":"The Babel Team (https://babel.dev/team)","license":"MIT","types":"./lib/index-legacy.d.ts"}
