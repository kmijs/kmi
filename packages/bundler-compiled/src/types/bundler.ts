import type { PluginItem as BabelPlugin } from '../../compiled/@babel/core'

export interface ICopy {
  from: string
  to: string
}

export type { BabelPlugin }

export enum CSSMinifier {
  esbuild = 'esbuild',
  cssnano = 'cssnano',
  lightningcss = 'lightningcss',
  none = 'none',
}

export enum JSMinifier {
  swc = 'swc',
  none = 'none',
  terser = 'terser',
  esbuild = 'esbuild',
}
