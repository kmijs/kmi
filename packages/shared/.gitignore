compiled/@ks/weblogger/**/*.d.ts
compiled/@ks/weblogger/es
compiled/@ks/weblogger/lib/types
compiled/@ks/weblogger/lib/log.bridge.js
compiled/@ks/weblogger/lib/log.browser.js
compiled/@ks/weblogger/lib/log.hybrid.js
compiled/@ks/weblogger/lib/log.kdsnative.js
compiled/@ks/weblogger/lib/log.krn.js
compiled/@ks/weblogger/lib/log.krnbridge.js
compiled/@ks/weblogger/lib/log.lite.js
compiled/@ks/weblogger/lib/log.mina.js
compiled/@ks/weblogger/lib/log.pure.js
compiled/@ks/weblogger/lib/plugins
compiled/@ks/weblogger/CHANGELOG.md
compiled/@ks/weblogger/README.md
docs
