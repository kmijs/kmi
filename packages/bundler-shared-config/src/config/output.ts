import { pathe } from '@kmijs/shared'

import { CHAIN_ID } from '@kmijs/bundler-shared'
import type { SharedConfigOptions } from '../types'
import { Env } from '../types'
import { getFilename } from '../utils/getFilename'
import { getOutputPath } from '../utils/getOutputPath'
export function applyOutput(opts: SharedConfigOptions) {
  const { config, userConfig, useHash, staticPathPrefix, bundler } = opts

  const isDev = opts.env === Env.development

  // output
  const absOutputPath = pathe.resolve(opts.cwd, getOutputPath(opts, 'root'))
  const jsPath = getOutputPath(opts, 'js')
  const jsFilename = getFilename(opts, 'js')
  const isJsFilenameFn = typeof jsFilename === 'function'

  const jsAsyncFilename = getFilename(opts, 'jsAsync')
  const isJsAsyncFilenameFn = typeof jsAsyncFilename === 'function'

  const disableCompress = process.env.COMPRESS === 'none'
  config.output
    .path(absOutputPath)
    .filename(
      isJsFilenameFn
        ? (...args) => {
            const name = jsFilename(...args)
            return pathe.join(jsPath, name)
          }
        : pathe.join(jsPath, jsFilename),
    )
    .chunkFilename(
      isJsAsyncFilenameFn
        ? (...args) => {
            const name = jsAsyncFilename(...args)
            return pathe.join(jsPath, name)
          }
        : pathe.join(jsPath, jsAsyncFilename),
    )
    .publicPath(userConfig.publicPath || 'auto')
    .pathinfo(isDev || disableCompress)
    .set(
      'assetModuleFilename',
      useHash
        ? `${staticPathPrefix}[name].[contenthash:8][ext]`
        : `${staticPathPrefix}[name][ext]`,
    )
    .set('hashFunction', 'xxhash64') // https://github.com/webpack/webpack/issues/14532#issuecomment-947525539

  // css output
  if (!userConfig.styleLoader) {
    const cssPath = getOutputPath(opts, 'css')
    const cssFilename = getFilename(opts, 'css')
    const cssAsyncFilename = getFilename(opts, 'cssAsync')

    const isCssFilenameFn = typeof cssFilename === 'function'
    const isCssAsyncFilenameFn = typeof cssAsyncFilename === 'function'

    config
      .plugin(CHAIN_ID.PLUGIN.MINI_CSS_EXTRACT)
      .use(bundler.CssExtractPlugin, [
        {
          filename: isCssFilenameFn
            ? (...args: any) => {
                // @ts-expect-error
                const name = cssFilename(...args)
                return pathe.join(cssPath, name)
              }
            : pathe.join(cssPath, cssFilename),
          chunkFilename: isCssAsyncFilenameFn
            ? (...args: any) => {
                // @ts-expect-error
                const name = cssAsyncFilename(...args)
                return pathe.join(cssPath, name)
              }
            : pathe.join(cssPath, cssAsyncFilename),
          ignoreOrder: true,
        },
      ])
  }
}
