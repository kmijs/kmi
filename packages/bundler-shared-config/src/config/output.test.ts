import Config from '@kmijs/bundler-shared/rspack-chain'
import { stringifyConfig } from '@kmijs/shared'
import { describe, expect, test } from 'vitest'
import { Env, type SharedConfigOptions } from '../types'
import { applyOutput } from './output'

class CssExtractPlugin {
  apply(compiler: any) {}
}

const createConfig = (opts: Partial<SharedConfigOptions>) => {
  const config = new Config()
  applyOutput({
    cwd: __dirname,
    env: Env.production,
    config,
    bundler: {
      CssExtractPlugin,
    },
    ...opts,
  } as SharedConfigOptions)
  return config.toConfig()
}
describe('output', () => {
  test('normal', () => {
    const config = createConfig({
      userConfig: {},
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('normal dev', () => {
    const config = createConfig({
      userConfig: {},
      staticPathPrefix: 'static/',
      env: Env.development,
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('应该允许使用 outputPath 自定义目录', () => {
    const config = createConfig({
      userConfig: {
        outputPath: 'dist/server',
      },
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('应该允许使用 outputPath.root 自定义目录', () => {
    const config = createConfig({
      userConfig: {
        outputPath: {
          root: 'dist/server',
        },
      },
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('should allow to set outputPath.js and outputPath.css to empty string', () => {
    const config = createConfig({
      userConfig: {
        outputPath: {
          js: '',
          css: '',
        },
      },
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('应该允许通过 outputPath.js 自定义 js 路径', () => {
    const config = createConfig({
      userConfig: {
        outputPath: {
          js: 'static/js',
        },
      },
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })

  test('允许 自定义 filename', () => {
    const config = createConfig({
      userConfig: {
        filename: {
          js: '[name].[contenthash:10].js',
          jsAsync: 'async/[name].[contenthash:8].js',
          css: '[name].[contenthash:10].css',
          cssAsync: 'async/[name].[contenthash:8].css',
        },
      },
      staticPathPrefix: 'static/',
    })
    expect(stringifyConfig(config)).toMatchSnapshot()
  })
})
