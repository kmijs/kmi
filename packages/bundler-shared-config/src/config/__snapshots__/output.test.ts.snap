// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`output > normal 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist',
    filename: '[name].js',
    chunkFilename: '[name].async.js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: '[name].css',
        chunkFilename: '[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > normal dev 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist',
    filename: '[name].js',
    chunkFilename: '[name].async.js',
    publicPath: 'auto',
    pathinfo: true,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: '[name].css',
        chunkFilename: '[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > should allow to set outputPath.js and outputPath.css to empty string 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist',
    filename: '[name].js',
    chunkFilename: '[name].async.js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: '[name].css',
        chunkFilename: '[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > 允许 自定义 filename 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist',
    filename: '[name].[contenthash:10].js',
    chunkFilename: 'async/[name].[contenthash:8].js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: '[name].[contenthash:10].css',
        chunkFilename: 'async/[name].[contenthash:8].css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > 应该允许使用 outputPath 自定义目录 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist/server',
    filename: '[name].js',
    chunkFilename: '[name].async.js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: '[name].css',
        chunkFilename: '[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > 应该允许使用 outputPath.root 自定义目录 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist/server',
    filename: 'js/[name].js',
    chunkFilename: 'js/[name].async.js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: 'css/[name].css',
        chunkFilename: 'css/[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;

exports[`output > 应该允许通过 outputPath.js 自定义 js 路径 1`] = `
"{
  output: {
    path: '<WORKSPACE>/packages/bundler-shared-config/src/config/dist',
    filename: 'static/js/[name].js',
    chunkFilename: 'static/js/[name].async.js',
    publicPath: 'auto',
    pathinfo: false,
    assetModuleFilename: 'static/[name][ext]',
    hashFunction: 'xxhash64'
  },
  plugins: [
    /* config.plugin('mini-css-extract-plugin') */
    new CssExtractPlugin(
      {
        filename: 'css/[name].css',
        chunkFilename: 'css/[name].async.css',
        ignoreOrder: true
      }
    )
  ]
}"
`;
